File: frontend/src/lib/eli/DokumentExpressionEli.ts
import { NormExpressionEli } from "@/lib/eli/NormExpressionEli"
import { NormWorkEli } from "@/lib/eli/NormWorkEli"

export const PATH_PARAMETER_SUBTYPE = `Subtype(regelungstext-verkuendung-[0-9]+|anlage-regelungstext-[0-9]+|bekanntmachungstext-[0-9]+|rechtsetzungsdokument-[0-9]+)`

/**
 * European legislation identifier on expression level for a Dokument of a Norm
 */
export class DokumentExpressionEli {
  readonly agent: string
  readonly year: string
  readonly naturalIdentifier: string
  readonly pointInTime: string
  readonly version: number
  readonly language: string
  readonly subtype: string

  constructor(
    agent: string,
    year: string,
    naturalIdentifier: string,
    pointInTime: string,
    version: number,
    language: string,
    subtype: string,
  ) {
    this.agent = agent
    this.year = year
    this.naturalIdentifier = naturalIdentifier
    this.pointInTime = pointInTime
    this.version = version
    this.language = language
    this.subtype = subtype
  }

  /**
   * Create an expression level ELI from a string representation
   *
   * @param eli the string representation of the ELI
   * @return the eli
   */
  static fromString(eli: string): DokumentExpressionEli {
    const match =
      /eli\/bund\/(?<agent>[^/]+)\/(?<year>[^/]+)\/(?<naturalIdentifier>[^/]+)\/(?<pointInTime>[^/]+)\/(?<version>[^/]+)\/(?<language>[^/]+)\/(?<subtype>[^/.]+)/.exec(
        eli,
      )

    if (match == null) {
      throw new Error("Invalid Dokument Expression Eli")
    }

    return new DokumentExpressionEli(
      match?.groups?.agent ?? "",
      match?.groups?.year ?? "",
      match?.groups?.naturalIdentifier ?? "",
      match?.groups?.pointInTime ?? "",
      parseInt(match?.groups?.version ?? "0", 10),
      match?.groups?.language ?? "",
      match?.groups?.subtype ?? "",
    )
  }

  /**
   * Create a DokumentExpressionEli from a NormExpressionEli string
   *
   * @param eli The NormExpressionEli string
   * @param subtype The document subtype (default: "regelungstext-verkuendung-1")
   * @return the DokumentExpressionEli
   */
  static fromNormExpressionEli(
    eli: NormExpressionEli,
    subtype: string = "regelungstext-verkuendung-1",
  ): DokumentExpressionEli {
    return new DokumentExpressionEli(
      eli.agent,
      eli.year,
      eli.naturalIdentifier,
      eli.pointInTime,
      eli.version,
      eli.language,
      subtype,
    )
  }

  /**
   * Create a NormExpressionEli that contains the parts of this ELI
   *
   * @return a norm ELI
   */
  asNormEli(): NormExpressionEli {
    return new NormExpressionEli(
      this.agent,
      this.year,
      this.naturalIdentifier,
      this.pointInTime,
      this.version,
      this.language,
    )
  }

  /**
   * Create a NormWorkEli that contains the parts of this ELI
   *
   * @return a norm ELI
   */
  asNormWorkEli(): NormWorkEli {
    return new NormWorkEli(this.agent, this.year, this.naturalIdentifier)
  }

  toString(): string {
    return `eli/bund/${this.agent}/${this.year}/${this.naturalIdentifier}/${this.pointInTime}/${this.version}/${this.language}/${this.subtype}`
  }

  equals(other: DokumentExpressionEli): boolean {
    if (other === this) return true
    else if (!(other instanceof DokumentExpressionEli)) return false
    else return other.toString() === this.toString()
  }
}

File: backend/src/main/java/de/bund/digitalservice/ris/norms/domain/entity/eli/DokumentExpressionEli.java
package de.bund.digitalservice.ris.norms.domain.entity.eli;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.Objects;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.jetbrains.annotations.NotNull;

/**
 * European legislation identifier on expression level for a dokument of a norm
 *
 * <p>This class can be used to extract the eli from a path that includes a section like
 * "/eli/bund/{agent}/{year}/{naturalIdentifier}/{pointInTime}/{version}/{language}/{subtype}".
 */
@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
public final class DokumentExpressionEli implements DokumentEli, Comparable<DokumentExpressionEli> {

  private String agent;
  private String year;
  private String naturalIdentifier;
  private LocalDate pointInTime;
  private Integer version;
  private String language;
  private String subtype;

  @Override
  public String toString() {
    return "eli/bund/%s/%s/%s/%s/%d/%s/%s".formatted(
      getAgent(),
      getYear(),
      getNaturalIdentifier(),
      getPointInTime().format(DateTimeFormatter.ISO_LOCAL_DATE),
      getVersion(),
      getLanguage(),
      getSubtype()
    );
  }

  /**
   * Create a {@link NormExpressionEli} that contains the parts of this eli
   *
   * @return a norm eli
   */
  @Override
  public NormExpressionEli asNormEli() {
    return new NormExpressionEli(
      getAgent(),
      getYear(),
      getNaturalIdentifier(),
      getPointInTime(),
      getVersion(),
      getLanguage()
    );
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    DokumentExpressionEli that = (DokumentExpressionEli) o;
    return Objects.equals(toString(), that.toString());
  }

  @Override
  public int hashCode() {
    return Objects.hash(toString());
  }

  @Override
  public int compareTo(@NotNull DokumentExpressionEli o) {
    return String.CASE_INSENSITIVE_ORDER.compare(this.toString(), o.toString());
  }
}
